### ISSUE TITLE
Failed requests queued

### ISSUE THREAD

**Issue by shravan2x** [aspnet/KestrelHttpServer#1209]
**Problem**:
Kestrel blocks all websocket requests for 10-20s seconds after a failed one.

**Reproduction**:
1. Create a websocket endpoint without `wss`.
2. Connect using `wss`.
3. Make two other connections using `ws`.
4. The two connections from (3) will block until (2) errors about 20s later.

**Log**:
These requests show up all together, suddenly.
```
info: Microsoft.AspNetCore.Server.Kestrel[17]
      Connection id "0HL09I0Q7I2S3" bad request data: "Malformed request: MethodIncomplete"
Microsoft.AspNetCore.Server.Kestrel.BadHttpRequestException: Malformed request: MethodIncomplete
info: Microsoft.AspNetCore.Hosting.Internal.WebHost[1]
      Request starting HTTP/1.1 GET http://127.0.0.1:5000/applebob/ws/control/?encoding=text
info: Microsoft.AspNetCore.Hosting.Internal.WebHost[1]
      Request starting HTTP/1.1 GET http://127.0.0.1:5000/applebob/ws/control/?encoding=text
```

**More info**:
My endpoint is designed to broadcast a message as soon as the connection is accepted. This should happen:
```
CONNECTED

RECEIVED: {"type":4,"connections":2}
```

But this happens:
```
CONNECTED

RECEIVED: {"type":4,"connections":2}

RECEIVED: {"type":4,"connections":1}
```

since both connections are accepted in quick succession after the failed one (times out?).

The connections were all made using https://www.websocket.org/echo.html

**Comment by cesarbs** [comment]
What version of Kestrel is this?

**Comment by Shravan2x** [comment]
The version is 1.0.1, the bundled one with .NET Core 1.1.0

**Comment by cesarbs** [comment]
It looks like you are hitting this issue: https://github.com/dotnet/corefx/issues/13369

We've recently checked in a workaround for it in Kestrel: https://github.com/aspnet/KestrelHttpServer/pull/1196

Can you try it with the latest development bits for Kestrel? Using feed `https://dotnet.myget.org/F/aspnetcore-dev/api/v3/index.json` should give you the right package.

**Comment by cesarbs** [comment]
cc @anurse

**Comment by Shravan2x** [comment]
My issue is on Windows though, that issue only mentions OSX and Linux. Same issue?

**Comment by cesarbs** [comment]
I just looked at the log message again and I see what the problem is: you're hitting and HTTP endpoint over a secure connection. You should be pointing your client to an encrypted endpoint.

The behavior you're seeing is due to Kestrel trying to parse your request as an HTTP request and spinning indefinitely because it can't find a space after the method. You client times out after 20 seconds and sends a FIN, causing Kestrel to error out saying the request method was incomplete (because there was no space after it).

**Comment by shravan2x** [comment]
I imagined it would be something like that, but why queue all following requests?

**Comment by cesarbs** [comment]
How many cores does your machine have?

**Comment by shravan2x** [comment]
4 cores, 8 threads. It's an i7-4790k.

**Comment by cesarbs** [comment]
How many failed requests are you seeing? They could all be timing out at once if they're concurrent connections.

Also, if for any reason you're setting the thread count to 1, that would explain the queuing.

**Comment by shravan2x** [comment]
Just one failed, the second two are valid and block until the first one times out.

I didn't change any of the defaults besides what VS2015 set.

**Comment by cesarbs** [comment]
Are you connecting to Kestrel directly or is it running behind IIS/IIS Express?

**Comment by shravan2x** [comment]
Directly.

**Comment by halter73** [comment]
Thanks for the report. A few more questions. 

- Do you see three unique connection ids in your log when you try this?
- Are you using the Microsoft.AspNetCore.WebSockets/0.2.0-preview1-final package? 
- What browser are you using?
- Can you repro with the app at https://github.com/aspnet/WebSockets/tree/dev/samples/EchoApp?

**Comment by shravan2x** [comment]
The myget link @cesarbs gave me (tps://dotnet.myget.org/F/aspnetcore-dev/api/v3/index.json) seems to have corrupted my solution in some way - `The schema version of 'Microsoft.AspNetCore.Server.IISIntegration.Tools' is incompatible with version 2.12.0.815 of NuGet. Please upgrade NuGet to the latest version from http://go.microsoft.com/fwlink/?LinkId=213942.`

I'll test the stuff you mentioned as soon as I figure this out.

**Comment by shravan2x** [comment]
Interesting, it looks like Chrome is queuing the connections, not Kestrel. Chrome appears to have a *one websocket connection attempt at a time* limit of something. Since Kestrel takes 20s to timeout, the other connections just wait on it.

Could we have Kestrel close failed connections immediately instead of waiting 20s?

**Comment by halter73** [comment]
@shravan2x I tested this with the latest bits using the app at, and it looks like the issue is fixed in the latest 1.1.0 packages. To address #1116, #1122 refactored request line validation which may have allowed Kestrel to reject the TLS request to a non-TLS endpoint more quickly.

In my observation, the request was rejected in far less than a second. According to the Chrome 54 network tools the request lengths were consistently less the 4ms when made to localhost.

@shravan2x If you try the newer packages, and feel this issue was closed in error just leave a comment and we can reopen.

![image](https://cloud.githubusercontent.com/assets/54385/20327300/58c6a994-ab41-11e6-8cef-940789d590ad.png)

![image](https://cloud.githubusercontent.com/assets/54385/20327312/610de784-ab41-11e6-94e1-71d064a230e6.png)

    C:\Users\shalter\dev\Universe\WebSockets\samples\EchoApp [dev = +1 ~0 -0 !]> dotnet run
    Project Microsoft.AspNetCore.WebSockets (.NETStandard,Version=v1.3) was previously compiled. Skipping compilation.
    Project EchoApp (.NETCoreApp,Version=v1.1) will be compiled because inputs were modified
    Compiling EchoApp for .NETCoreApp,Version=v1.1
    C:\Users\shalter\dev\Universe\WebSockets\samples\EchoApp\project.json(7,48): warning NU1007: Dependency specified was Microsoft.AspNetCore.WebSockets >= 1.0.0-* but ended up with Microsoft.AspNetCore.WebSockets 1.1.0.

    Compilation succeeded.
        1 Warning(s)
        0 Error(s)

    Time elapsed 00:00:04.9927451


    dbug: Microsoft.AspNetCore.Hosting.Internal.WebHost[3]
          Hosting starting
    dbug: Microsoft.AspNetCore.Hosting.Internal.WebHost[4]
          Hosting started
    Hosting environment: Production
    Content root path: C:\Users\shalter\dev\Universe\WebSockets\samples\EchoApp
    Now listening on: http://localhost:5000
    Application started. Press Ctrl+C to shut down.
    dbug: Microsoft.AspNetCore.Server.Kestrel[1]
          Connection id "0HL0DL02IQHI0" started.
    info: Microsoft.AspNetCore.Server.Kestrel[17]
          Connection id "0HL0DL02IQHI0" bad request data: "Invalid request line: <0x16><0x03><0x01><0x00><0xA7><0x01><0x00><0x00><0xA3><0x03><0x03><0xE5><0xE4><0x97>p<0x13>wY<0xAB><0x8B><0xE9><0xA1>J&<0x00>lvk<0xBD><0xBF><0xC6>N..."
    Microsoft.AspNetCore.Server.Kestrel.BadHttpRequestException: Invalid request line: <0x16><0x03><0x01><0x00><0xA7><0x01><0x00><0x00><0xA3><0x03><0x03><0xE5><0xE4><0x97>p<0x13>wY<0xAB><0x8B><0xE9><0xA1>J&<0x00>lvk<0xBD><0xBF><0xC6>N...
    dbug: Microsoft.AspNetCore.Server.Kestrel[10]
          Connection id "0HL0DL02IQHI0" disconnecting.
    dbug: Microsoft.AspNetCore.Server.Kestrel[7]
          Connection id "0HL0DL02IQHI0" sending FIN.
    dbug: Microsoft.AspNetCore.Server.Kestrel[8]
          Connection id "0HL0DL02IQHI0" sent FIN with status "0".
    dbug: Microsoft.AspNetCore.Server.Kestrel[2]
          Connection id "0HL0DL02IQHI0" stopped.
    dbug: Microsoft.AspNetCore.Server.Kestrel[1]
          Connection id "0HL0DL02IQHI1" started.
    info: Microsoft.AspNetCore.Server.Kestrel[17]
          Connection id "0HL0DL02IQHI1" bad request data: "Invalid request line: <0x16><0x03><0x01><0x00><0xA7><0x01><0x00><0x00><0xA3><0x03><0x03>K<0xD9><0xCC>@/;<0xBF><0xA4><0xD6><0x8A>.S2A<0xE9>_<0xD0>J<0xA8>1<0x16>..."
    Microsoft.AspNetCore.Server.Kestrel.BadHttpRequestException: Invalid request line: <0x16><0x03><0x01><0x00><0xA7><0x01><0x00><0x00><0xA3><0x03><0x03>K<0xD9><0xCC>@/;<0xBF><0xA4><0xD6><0x8A>.S2A<0xE9>_<0xD0>J<0xA8>1<0x16>...
    dbug: Microsoft.AspNetCore.Server.Kestrel[10]
          Connection id "0HL0DL02IQHI1" disconnecting.
    dbug: Microsoft.AspNetCore.Server.Kestrel[6]
          Connection id "0HL0DL02IQHI1" received FIN.
    dbug: Microsoft.AspNetCore.Server.Kestrel[7]
          Connection id "0HL0DL02IQHI1" sending FIN.
    dbug: Microsoft.AspNetCore.Server.Kestrel[8]
          Connection id "0HL0DL02IQHI1" sent FIN with status "0".
    dbug: Microsoft.AspNetCore.Server.Kestrel[2]
          Connection id "0HL0DL02IQHI1" stopped.

**Comment by shravan2x** [comment]
Yep that did it. Thanks!
